src/lib/components/Swap/SwapButton/useApprovalData.tsx

```diff
@@ -32,61 +32,57 @@ export default function useApprovalData(
     currencyAmount
   )
 
+  const [isPending, setIsPending] = useState(false)
   const addTransaction = useAddTransaction()
   const onApprove = useCallback(async () => {
+    setIsPending(true)
     const transaction = await handleApproveOrPermit()
     if (transaction) {
       addTransaction({ type: TransactionType.APPROVAL, ...transaction })
     }
+    setIsPending(false)
   }, [addTransaction, handleApproveOrPermit])
+  // Reset the pending state if currency changes.
+  useEffect(() => setIsPending(false), [currency])
 
   const approvalHash = usePendingApproval(currency?.isToken ? currency : undefined, useSwapRouterAddress(trade))
-  const approvalData = useMemo((): Partial<ActionButtonProps> | undefined => {
+  const approvalAction = useMemo((): Action | undefined => {
     if (!trade || !currency) return
 
-    if (approvalState === ApproveOrPermitState.REQUIRES_APPROVAL) {
-      return {
-        action: {
+    switch (approvalState) {
```

Comment:
Note that only returning from within the `switch` statement ensures that all cases are covered, as the compiler will err if there is a codepath with no return.